specs/ServerCertificate.md

```diff
@@ -0,0 +1,264 @@
+                if (error != S_OK)
+                {
+                  ShowFailure( error,  L"Clear server certificate error override cache failed");
+                }
+                MessageBox(nullptr, L"Cleared", L"Clear server certificate error override cache", MB_OK);
+                return S_OK;
+            })
+            .Get()));
+}
+```
+## . NET/ WinRT
+```c#
+// This example bypasses the default TLS interstitial page using the ReceivingServerCertificateError
+// event handler and continues with the navigation to a server with a TLS certificate that is signed by
+// an authority that WebView2 doesn't trust. Otherwise, cancel the request.
+private bool _isServerCertificateError = false;
+void EnableCustomServerCertificateError()
+{
+  // Safeguarding the handler when unsupported runtime is used.
+  try
+  {
+    if (!_isServerCertificateError)
+    {
+      webView.CoreWebView2.ReceivingServerCertificateError += WebView_ReceivingServerCertificateError;
+    }
+    else
+    {
+      webView.CoreWebView2.ReceivingServerCertificateError -= WebView_ReceivingServerCertificateError;
+    }
+    _isServerCertificateError = !_isServerCertificateError;
+
+    MessageBox.Show(this,
+      _isServerCertificateError ? "Custom server certificate error has been enabled" : "Custom server certificate error has been disabled",
+      "Custom server certificate error");
+  }
+  catch (NotImplementedException exception)
+  {
+    MessageBox.Show(this, "Custom server certificate error Failed: " + exception.Message, "Custom server certificate error");
+  }
+}
+
+void WebView_ReceivingServerCertificateError(object sender, CoreWebView2ReceivingServerCertificateErrorEventArgs e)
+{
+  CoreWebView2Certificate certificate = e.ServerCertificate;
+
+  // Continues with the navigation to a server with a TLS certificate if
+  // the certificate is invalid and signed by badssl authority.
+  if (e.ErrorStatus == CoreWebView2WebErrorStatus.CertificateIsInvalid &&
+      (certificate.Issuer == "*.badssl.com" ||
+      certificate.Issuer == "BadSSL Untrusted Root Certificate Authority"))
+  {
+    // Continue the request with the TLS certificate.
+    e.Handled = true;
+  }
+  else
+  {
+    // Cancel the request.
+    e.Cancel = true;
+  }
+}
+
+// This example clears the TLS decision in response to proceeding with TLS certificate errors.
+void ClearServerCertificateErrorOverrideCache()
+{
+  await webView.CoreWebView2.ClearServerCertificateErrorOverrideCacheAsync();
+  MessageBox.Show(this, "Cleared", "Clear server certificate error override cache");
+}
+```
+
+# API Details
+
+## Win32 C++
+``` cpp
+[uuid(4B7FF0D2-8203-48B0-ACBF-ED9CFF82567A), object, pointer_default(unique)]
+interface ICoreWebView2_11 : ICoreWebView2_10 {
+  /// Add an event handler for the ServerCertificateError event.
+  /// The ServerCertificateError event is raised when the WebView2
+  /// cannot verify server's digital certificate while loading a web page.
```

Comment:
>  while loading a web page.

This applies to more than just loading a web page correct? Sub-resources (images, script, and so on) or anything that goes through WebResourceRequested could have an invalid cert and would raise this event? Or is it only for navigations?

Reply:
Good question, haven't thought about this.

Looks like from my implementation, I am blocking/cancelling when a request is not a main frame/prefetch requests. And  for fetch/XMLHttpRequest the event gets raised.

In WebView2, do we want to raise event for all sub-resources for a cert failure and let developer decide want they wanna do?

Theoretically they would cancel the sub resource request anyways except (may/may not be)for xhr requests.

Reply:
> In WebView2, do we want to raise event for all sub-resources for a cert failure and let developer decide want they wanna do?

This makes the most sense to me. If a WebView2 starts on goodtls.com/index.html and that page has an img src=badtls.com/img.png and a link to badtls.com/index.html and they have the event handler hooked up to allow badtls.com, the image won't work until the end user clicks the link to badtls.com/index.html.

Reply:
👍